snuba/migrations/snuba_migrations/functions/0001_functions.py

```diff
@@ -0,0 +1,225 @@
+from typing import List, Sequence
+
+from snuba.clickhouse.columns import (
+    UUID,
+    AggregateFunction,
+    Array,
+    Column,
+    DateTime,
+    Float,
+    Nested,
+    String,
+    UInt,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.storages.tags_hash_map import TAGS_HASH_MAP_COLUMN
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+
+common_columns: List[Column[Modifiers]] = [
+    Column("project_id", UInt(64)),
+    Column("transaction_name", String()),
+    Column("timestamp", DateTime()),
+    Column("depth", UInt(32)),
+    Column("parent_fingerprint", UInt(64)),
+    Column("fingerprint", UInt(64)),
+    Column("symbol", String()),
+    Column("image", String()),
+    Column("filename", String()),
+    Column("is_application", UInt(8)),
+    Column("tags", Nested([Column("key", String()), Column("value", String())])),
+    Column("retention_days", UInt(16)),
+]
+
+raw_columns: List[Column[Modifiers]] = common_columns + [
+    Column("durations", Array(Float(64))),
+    Column("profile_id", UUID()),
+    Column("materialization_version", UInt(8)),
+]
+
+agg_columns: List[Column[Modifiers]] = common_columns + [
+    Column("count", AggregateFunction("count", [Float(64)])),
+    Column(
+        "percentiles",
+        AggregateFunction("quantiles(0.5, 0.75, 0.9, 0.95, 0.99)", [Float(64)]),
+    ),
+    Column("min", AggregateFunction("min", [Float(64)])),
+    Column("max", AggregateFunction("max", [Float(64)])),
+    Column("avg", AggregateFunction("avg", [Float(64)])),
+    Column("sum", AggregateFunction("sum", [Float(64)])),
+    Column("worst", AggregateFunction("argMax", [UUID(), Float(64)])),
+    Column("examples", AggregateFunction("groupUniqArray(5)", [UUID()])),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+    index_granularity = "2048"
+    storage_set = StorageSetKey.FUNCTIONS
+
+    data_granularity = 60 * 60  # 1 hour buckets
+
+    local_raw_table = "functions_raw_local"
+    dist_raw_table = "functions_raw_dist"
+
+    local_materialized_table = "functions_mv_local"
+    dist_materialized_table = "functions_mv_dist"
+
+    local_view_table = "functions_local"
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=self.storage_set,
+                table_name=self.local_raw_table,
+                columns=raw_columns,
+                engine=table_engines.MergeTree(
+                    storage_set=self.storage_set,
+                    order_by="(project_id, transaction_name, timestamp)",
+                    partition_by="(toStartOfInterval(timestamp, INTERVAL 12 HOUR))",
+                    ttl="timestamp + toIntervalDay(1)",
```

Comment:
Are you allowing backdated profiles in the DB? 
Like would you accept a profile with a timestamp older than 24 hours ago? 
If yes here you may have a problem and you should verify that. If you insert a row with a timestamp older than 24 hours, does it get materialized or is it dropped immediately ?

Reply:
For now, this should be impossible because profiles are dated with the timestamp they were received. But I have verified if a row is inserted with a timestamp older than 24 hours, it's not guaranteed to be dropped immediately. It looks like it'll automatically delete rows past its TTL but clickhouse doesn't guarantee when that'll happen, so we probably should not rely on this behaviour if we want to accept backdated profiles.